accepted/future-releases/language-versioning/package-config-file-v2.md

```diff
@@ -0,0 +1,193 @@
+
+Any tool which modifies the file piecemeally by changing or add information, rather than writing it from scratch like Pub, should retain all properties that it does not recognize. This allows multiple tools to all cooperate about adding information to the file. 
+
+Again, Pub is not required to retain any information, or read the file at all, it can overwrite the file when invoked.
+
+#### Package Names
+
+Some properties are specified as being "package names".
+
+A package name is used as both a URI path segment and a directory name. To ensure this is possible, we define a package name as a string containing:
+
+- Only [URI path characters](https://www.ietf.org/rfc/rfc3986.txt) (RFC 3986 "pchar"),
+- but no `:` or `%`,
+- and containing at least one non-`.` character.
+
+This allows the characters: `a`-`z`, `A`-`Z`, `0`-`9`, `-`, `.`, `_`, `~`, `!`, `$`, `&`, `'`, `(`, `)`, `*`, `+`, `,`, `;`, `=`, `@`. 
+It avoids the characters `/`, `\` and `:` which are meaningful in paths on various operating systems, as well as the strings `.` and `..`, and it ensures that the package name is a valid non-empty URI path segment.
+
+Pub has stronger restrictions on its accepted package names, but this file will also be used in settings where packages are not supplied by Pub.
+
+### Finding the File
+
+When compiling a Dart script file, the Dart tool will look for the necessary information starting in the directory of the script file. It will first check for a `.dart_tool/package_config.json` file, then for a `.packages` file, the it will repeat this through the parent directories until it one such file or it reaches the file system root.
+
+The analyzer needs to cleverly determine which configuration file applies to which file, based on the location of scripts, which scripts may transitively import a file, and understanding of the structure of Pub packages. It should be able to use the same algorithm as now, just that where it currently looks for a `.packages` in a directory, it will have to for `.dart_tool/package_config.json` first.
+
+### Specifying the File
+
+Tools currently allow you to specify the `.packages` file with a command line argument:
+
+```
+--packages=someFile
+-p someFile
+```
+
+This approach will work with the JSON configuration file as well. The tool must scan the file to detect whether it is a JSON file or an original `.packages` file. It is considered a JSON file if the first non-whitespace (space, tab, carriage return, linefeed) character is a `{`. The JSON file must start this way, and the existing format cannot do so; it must start with either a comment, `#`, or a non-empty package name, and package names cannot contain a `{` character.
```

Comment:
What do you think about allowing tools to look for a .dart_tools/package_config.json file in the same directory as the specified .packages file and use that instead?

Reply:
Alternatively, what about having tools (e.g. pub) which are generating both files mark the .packages file with a redirection marker so that other tools which are aware of both conventions can redirect to the non-deprecated version if they are pointed at the deprecated version?  This could possibly go in the initial comment if we wanted to make it completely non-breaking.

Reply:
I believe there was widespread support for keeping the file out of the package root directory at the meeting where the location was decided. We can allow you to put it there, but then we'd have to decide which of `package_config.json` and `.dart_tool/package_config.json` takes precedence. If Pub is always going to generate the file in `.dart_tool`, then I'm not sure there is an advantage in also allowing it elsewhere.

I'm not sure which problem the redirection is trying to solve. Is it the case where a tool is explicitly pointed to the `.packages` file, fx with a command line argument like `-p .packages`, and then it figures out to use the `.dart_tool/package_config.json` instead because of the redirect?

We could make our configuration-file-finding-algorithm automatically check for `.dart_tool/package_config.json` when it detects that it's being asked to load a `.packages` file. It's currently forced to detect the content by pattern matching the syntax (luckily it can be seen from the first character).

Reply:
Sorry, I wasn't proposing to change the location.  The issue I'm trying to solve is that there seem to be many (external) tools that pass the location of a .packages file explicitly to another tool.  For example, I believe that many of the uses in Flutter that @jakemac53 found are in this category: they neither generate nor parse the contents of .packages, they just say "here's the .packages file!".  If the eventual tool which does parse the contents has already been updated, it seems like it might be valuable for that tool to be able to say "Ah, you have pointed me at a .packages file, but I see that there is a .dart_tool/package_config.json file in the same directory, so I will just use that instead".  But if that is considered too surprising or error prone, then we could instead only do that if the .packages file explicitly says "Tools that have been updated should use this other file, otherwise here's the legacy contents".  Hence the alternative proposal for a redirection.

Reply:
Added an "if you are pointed to a .packages file, you *should* check for an adjacent .dart_tool/package_config.json file" clause.